Transform "documentation" tab into button

GuillaumeGomez opened the pull request:
Same as before but with a fix for the 404 pages and a test to avoid regressions.

EDIT: Back to tab style but the "documentation" tab is now a button:

mobile:

![Screenshot from 2020-09-08 17-50-40](https://user-images.githubusercontent.com/3050060/92499385-171c3a80-f1fc-11ea-8189-4e542988801f.png)

"normal":

![Screenshot from 2020-09-08 17-50-37](https://user-images.githubusercontent.com/3050060/92499391-17b4d100-f1fc-11ea-9cd7-a7a52e2a3f9e.png)

jyn514 commented:
Now that I think about this more, I'm not sure about this change ... it makes the /crate page very inconsistent with respect to the release pages, e.g. https://docs.rs/releases/stars.

GuillaumeGomez commented:
It's voluntary and I proposed to discussed about it. Discussion is still open. :)

pietroalbini commented:
Yeah, I think chasing the rustdoc styling for docs.rs is not the best choice, especially if we want to eventually make docs.rs in line with the website styling.

GuillaumeGomez commented:
Agreed, that's why when we're still in "crate focus", we have a sidebar simply because the tab system is just not adapted. However, for the release pages, it totally makes sense (in my opinion).

GuillaumeGomez commented:
Except that the current tab style is not appropriate for the crate pages because the behavior of the tabs isn't consistent.

pietroalbini commented:
I think having the button be more visible should help:

![2020-09-07--13-09-40](https://user-images.githubusercontent.com/2299951/92381689-9847d480-f10b-11ea-924b-4653648ad026.png)

pietroalbini commented:
Looks good! Two notes:

* Let's add some margin between the description and the button, to avoid having the ellipsis so close to the button itself.
* I'd add a bit more padding to the button (dunno, `10px 15px`?).

pietroalbini commented:
Also, it seems a bit off-centered.

GuillaumeGomez commented:
> Also, it seems a bit off-centered.

A bit more of precision on this one please. :laughing: (I'll implement the other remarks in the meantime)

pietroalbini commented:
![2020-09-07--15-35-35](https://user-images.githubusercontent.com/2299951/92393329-05fdfb80-f120-11ea-8039-5c21333550df.png)

I think the best solution to fix this is to put everything in a flexbox (not just the description) and have all the contents of the header on the left and the button on the right-

pietroalbini commented:
Mostly looks good now! The only two changes I'd make:

* There are merge conflicts :)
* There should be some `padding-bottom` on mobile below the crate name, ideally the same as the `padding-top`. Currently it's too close to the tabs.

GuillaumeGomez commented:
Done!

jyn514 commented:
I'm unsure about this change too :P I like making 'Documentation' a separate button, because it takes you to a whole new page. I don't like though that it's over on the right side of the screen - it's harder to find if you're looking in the old place and it's also away from the main focus, which is the crate title.

Maybe we could put it in the center instead (literally front and center ;)? Then it would be easier to find.

GuillaumeGomez commented:
It's quite difficult to miss it though. But just by curiosity, where would you put it? (screenshot done with gimp please!)

jyn514 commented on src/web/error.rs:
```diff
@@ -93,3 +93,33 @@ impl From<PoolError> for IronError {
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```
```suggestion
```

jyn514 commented on src/web/releases.rs:
```diff
@@ -1213,4 +1213,57 @@ mod tests {
     }
+
+    #[test]
+    fn check_releases_page_content() {
```
Isn't there already a test for this?

GuillaumeGomez replied:
I thought to but couldn't find it.

jyn514 replied:
Here it is: https://github.com/rust-lang/docs.rs/blob/d688dc78f5d6e63f07b7c40b275e77817ef466c5/src/web/releases.rs#L1024
So I think I would prefer to remove this duplicate test.

GuillaumeGomez replied:
It doesn't test the same things. My test added checks specifically to ensure the links were not removed by mistake (which I did while changing the templates when working on this PR).

jyn514 reviewed (commented):
This changed the 'about' page to look rather strange, I think I prefered it before:
![image](https://user-images.githubusercontent.com/23638587/92479542-aa7c4e00-f1b1-11ea-88c2-c10fcfffde3c.png)

GuillaumeGomez commented:
> This changed the 'about' page to look rather strange, I think I prefered it before:

Then it's a bug! I'll fix it.

jyn514 commented on templates/style/base.scss:
```diff
@@ -444,87 +444,118 @@ div.cratesfyi-package-container {
-        }
-    }
+        .description-container {
+            flex-grow: 3;
```
Without this it looks a lot better IMO, the 'Documentation' tab is centered:
![image](https://user-images.githubusercontent.com/23638587/92479868-1d85c480-f1b2-11ea-83f6-8724ebdce776.png)

pietroalbini replied:
I'm pretty sure this way the button will jump around the page if the description is longer.

Nemo157 replied:
An example of a longer description: <https://docs.rs/crate/tokio/0.2.22>

jyn514 replied:
This is what it looks like on mobile: 
![image](https://user-images.githubusercontent.com/23638587/92480208-96851c00-f1b2-11ea-9c9c-01283a816961.png)

jyn514 replied:
> I'm pretty sure this way the button will jump around the page if the description is longer.

Hmm, maybe we can limit the length of the description and show this button after?

GuillaumeGomez replied:
It's kind of the case. I just don't like the fact that the button position might change, which is why I added `flex-grow: 3` so like that, whatever the content length, it remains on the right.

jyn514 replied:
Ok yeah, with `width: 50%;` it looks consistent:
![image](https://user-images.githubusercontent.com/23638587/92481322-2f686700-f1b4-11ea-8686-702f3082ee3d.png)
![image](https://user-images.githubusercontent.com/23638587/92481327-342d1b00-f1b4-11ea-8467-17b56e20f124.png)

jyn514 replied:
I actually prefer it with the text limited as well, it makes the lines easier to read when they're shorter.

GuillaumeGomez replied:
Fine by me. We'll have both then! :)

pietroalbini replied:
It feels... weird. Dunno why, but I'm not sure it's the right approach :/

GuillaumeGomez replied:
I don't like it either. Didn't add the line yet so feel free to debate about it. :)

jyn514 replied:
We decided to instead keep the right position, but make the button more noticable by adding a black background.

jyn514 commented:
About is fixed: 
![image](https://user-images.githubusercontent.com/23638587/92671181-98023180-f2e3-11ea-9c85-4e76d0dad026.png)
But releases are still broken: 
![image](https://user-images.githubusercontent.com/23638587/92671195-9f293f80-f2e3-11ea-9f6a-18e21098287d.png)

We should really think about #996 :/

I'll review the rest in the meantime so you can fix bugs in parallel ;)

jyn514 commented on src/web/releases.rs:
```diff
@@ -1213,4 +1213,57 @@ mod tests {
+                let page = kuchiki::parse_html()
+                    .one(env.frontend().get(url).send().unwrap().text().unwrap());
+                assert_eq!(page.select("#crate-title").unwrap().count(), 1);
+                let not_maching = page
```
nit: `not_matching`

jyn514 commented on src/web/releases.rs:
```diff
@@ -1213,4 +1213,57 @@ mod tests {
+
+    #[test]
+    fn check_releases_page_content() {
+        let sel = ".pure-menu-horizontal>.pure-menu-list>.pure-menu-item>.pure-menu-link>.title";
```
Hmm ... this seems a little fragile but I'm not sure how else to do it. I guess we're the only ones updating the releases page so whoever modifies the HTML can be in charge of fixing the test, but if so there should be a comment:
```suggestion
        // NOTE: this is a little fragile and may have to be updated if the HTML layout changes
        let sel = ".pure-menu-horizontal>.pure-menu-list>.pure-menu-item>.pure-menu-link>.title";
```

jyn514 commented on src/web/releases.rs:
```diff
@@ -1213,4 +1213,57 @@ mod tests {
+                    )
+                    .filter(|(a, b)| a.as_str() != **b)
+                    .collect::<Vec<_>>();
+                if !not_maching.is_empty() {
```
nit: double negatives are hard to read
```suggestion
                if not_maching.len() > 0 {
```

jyn514 commented on src/web/releases.rs:
```diff
@@ -1213,4 +1213,57 @@ mod tests {
+                    let found = not_maching.iter().map(|(a, _)| a).collect::<Vec<_>>();
+                    assert!(
+                        not_maching.is_empty(),
+                        "Titles not found for URL `{}`: {:?}, found: {:?}",
```
```suggestion
                        "Titles did not match for URL `{}`: not found: {:?}, found: {:?}",
```

jyn514 commented on templates/about-base.html:
```diff
@@ -1,32 +1,34 @@
             </div>
         </div>
-	</div>
+    </div>
```
Why is this changing `about-base` at all? Shouldn't this only need to modify /crate pages?

GuillaumeGomez replied:
Oh it was just a tab so I guess it was done by my IDE...

jyn514 replied:
No, it added `             <div class="description-container">` above, which doesn't seem right.

jyn514 commented on templates/style/base.scss:
```diff
@@ -444,87 +444,119 @@ div.cratesfyi-package-container {
-    h1 {
-        margin: 0;
-        padding: 15px 0 0 14px;
+    .container {
+        display: flex;
+        align-items: center;
```
I don't feel comfortable reviewing all this CSS, but since you and pietro worked it out I trust your judgement. The end result looks good :)

GuillaumeGomez commented:
Fixed release page as well and fixed the nits.

> We should really think about #996 :/

Yep but it's a tricky thing to do... We can use the same system we use in rustdoc but we still have the issue with getting the screenshot differences. But in here, I think this problem would be very limited considering that the build is quite small in comparison.

jyn514 commented:
Thanks for working on this!
